File: packages/ui-compose/ui/src/test/kotlin/com/kinetixui/ui/InteractionTest.kt
package com.kinetixui.ui

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.material3.Text
import androidx.compose.ui.state.ToggleableState
import androidx.compose.ui.test.junit4.v2.createComposeRule
import androidx.compose.ui.test.isToggleable
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/**
 * Interaction tests — the core controls do what their API says when a user acts on them. `assembleDebug` proves
 * they compile; these prove they behave: a press calls back once, a disabled control ignores input, a toggle
 * reports the new value.
 */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34])
class InteractionTest {
    @get:Rule
    val rule = createComposeRule()

    @Test
    fun button_click_invokes_onClick_once() {
        var clicks = 0
        rule.setContent { KinetixTheme(darkTheme = false) { KinetixButton(onClick = { clicks++ }) { Text("Save") } } }
        rule.onNodeWithText("Save").performClick()
        assertEquals(1, clicks)
    }

    @Test
    fun disabled_button_ignores_clicks() {
        var clicks = 0
        rule.setContent {
            KinetixTheme(darkTheme = false) { KinetixButton(onClick = { clicks++ }, enabled = false) { Text("Save") } }
        }
        rule.onNodeWithText("Save").performClick()
        assertEquals(0, clicks)
    }

    @Test
    fun every_button_variant_is_clickable() {
        val clicks = mutableMapOf<String, Int>()
        rule.setContent {
            KinetixTheme(darkTheme = false) {
                Column {
                    for (variant in KinetixButtonVariant.values()) {
                        KinetixButton(onClick = { clicks[variant.name] = (clicks[variant.name] ?: 0) + 1 }, variant = variant) {
                            Text(variant.name)
                        }
                    }
                }
            }
        }
        for (variant in KinetixButtonVariant.values()) {
            rule.onNodeWithText(variant.name).performClick()
            assertEquals("variant ${variant.name}", 1, clicks[variant.name])
        }
    }

    @Test
    fun checkbox_click_invokes_onClick() {
        var clicks = 0
        rule.setContent {
            KinetixTheme(darkTheme = false) { KinetixCheckbox(state = ToggleableState.Off, onClick = { clicks++ }) }
        }
        rule.onNode(isToggleable()).performClick()
        assertEquals(1, clicks)
    }

    @Test
    fun switch_reports_the_opposite_of_its_current_value() {
        var reported: Boolean? = null
        rule.setContent {
            KinetixTheme(darkTheme = false) { KinetixSwitch(checked = false, onCheckedChange = { reported = it }) }
        }
        rule.onNode(isToggleable()).performClick()
        assertEquals(true, reported)
    }

    @Test
    fun disabled_switch_ignores_clicks() {
        var reported: Boolean? = null
        rule.setContent {
            KinetixTheme(darkTheme = false) {
                KinetixSwitch(checked = false, onCheckedChange = { reported = it }, enabled = false)
            }
        }
        rule.onNode(isToggleable()).performClick()
        assertEquals(null, reported)
    }

    @Test
    fun toggle_reports_the_opposite_of_its_current_value() {
        var reported: Boolean? = null
        rule.setContent {
            KinetixTheme(darkTheme = false) {
                KinetixToggle(pressed = false, onPressedChange = { reported = it }) { Text("B") }
            }
        }
        rule.onNodeWithText("B").performClick()
        assertTrue(reported == true)
    }

    @Test
    fun tab_click_invokes_onClick() {
        var clicks = 0
        rule.setContent {
            KinetixTheme(darkTheme = false) {
                Row { KinetixTabsTrigger(text = "One", selected = false, onClick = { clicks++ }) }
            }
        }
        rule.onNodeWithText("One").performClick()
        assertEquals(1, clicks)
        assertFalse(clicks == 0)
    }
}

File: packages/ui-flutter/test/interaction_test.dart
// Interaction tests — the core controls do what their API says when a user acts on them.
// Smoke tests prove a widget *builds*; these prove it *behaves*: a press calls back once, a
// disabled control ignores input, a toggle reports the new value, typed text reaches `onChanged`.

// ignore_for_file: prefer_const_constructors, prefer_const_literals_to_create_immutables

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:kinetix_ui/kinetix_ui.dart';

Widget host(Widget child) => MaterialApp(
      home: KinetixTheme(
        brightness: Brightness.light,
        child: Scaffold(body: Center(child: child)),
      ),
    );

// kx-verify: interaction
void main() {
  group('KinetixButton', () {
    testWidgets('a tap calls onPressed once', (tester) async {
      var presses = 0;
      await tester.pumpWidget(host(KinetixButton(onPressed: () => presses++, child: const Text('Save'))));
      await tester.tap(find.byType(KinetixButton));
      await tester.pump();
      expect(presses, 1);
    });

    testWidgets('a null onPressed is disabled and ignores taps', (tester) async {
      await tester.pumpWidget(host(const KinetixButton(onPressed: null, child: Text('Save'))));
      await tester.tap(find.byType(KinetixButton));
      await tester.pump();
      expect(tester.takeException(), isNull);
    });

    for (final variant in KinetixButtonVariant.values) {
      testWidgets('${variant.name} variant is tappable', (tester) async {
        var presses = 0;
        await tester.pumpWidget(
          host(KinetixButton(variant: variant, onPressed: () => presses++, child: Text(variant.name))),
        );
        await tester.tap(find.byType(KinetixButton));
        expect(presses, 1);
      });
    }
  });

  group('KinetixCheckbox', () {
    testWidgets('a tap reports the opposite of the current value', (tester) async {
      bool? reported;
      await tester.pumpWidget(host(KinetixCheckbox(value: false, onChanged: (v) => reported = v)));
      await tester.tap(find.byType(KinetixCheckbox));
      expect(reported, isTrue);

      await tester.pumpWidget(host(KinetixCheckbox(value: true, onChanged: (v) => reported = v)));
      await tester.tap(find.byType(KinetixCheckbox));
      expect(reported, isFalse);
    });

    testWidgets('a null onChanged is disabled', (tester) async {
      await tester.pumpWidget(host(const KinetixCheckbox(value: false, onChanged: null)));
      await tester.tap(find.byType(KinetixCheckbox));
      expect(tester.takeException(), isNull);
    });
  });

  group('KinetixSwitch', () {
    testWidgets('a tap reports the opposite of the current value', (tester) async {
      bool? reported;
      await tester.pumpWidget(host(KinetixSwitch(value: false, onChanged: (v) => reported = v)));
      await tester.tap(find.byType(KinetixSwitch));
      expect(reported, isTrue);
    });

    testWidgets('a null onChanged is disabled', (tester) async {
      await tester.pumpWidget(host(const KinetixSwitch(value: true, onChanged: null)));
      await tester.tap(find.byType(KinetixSwitch));
      expect(tester.takeException(), isNull);
    });
  });

  group('KinetixToggle', () {
    testWidgets('a tap reports the opposite of the current value', (tester) async {
      bool? reported;
      await tester.pumpWidget(
        host(KinetixToggle(pressed: false, onChanged: (v) => reported = v, child: const Text('B'))),
      );
      await tester.tap(find.byType(KinetixToggle));
      expect(reported, isTrue);
    });
  });

  group('KinetixTabsTrigger', () {
    testWidgets('a tap calls onTap', (tester) async {
      var taps = 0;
      await tester.pumpWidget(host(KinetixTabsTrigger('One', selected: false, onTap: () => taps++)));
      await tester.tap(find.byType(KinetixTabsTrigger));
      expect(taps, 1);
    });
  });

  group('KinetixInput', () {
    testWidgets('typed text reaches onChanged', (tester) async {
      String? text;
      await tester.pumpWidget(host(SizedBox(width: 300, child: KinetixInput(onChanged: (v) => text = v))));
      await tester.enterText(find.byType(EditableText), 'hello');
      expect(text, 'hello');
    });
  });
}
